### PR TITLE
Add `test-title-format` rule

### LINKED ISSUE #157: Rule proposal: `test-title-format`
[Issuehunt badges](https://issuehunt.io/r/avajs/eslint-plugin-ava/issues/157)

Give to check the title on a specific regex pattern.
We adhere to specific naming titles, but the new members in the team are sometimes overlooked or do mistake.

> There is a $60.00 open bounty on this issue. Add more on [Issuehunt](https://issuehunt.io/r/avajs/eslint-plugin-ava/issues/157).
> - Checkout the [Issuehunt explorer](https://issuehunt.io/r/avajs/eslint-plugin-ava/) to discover more funded issues.
> - Need some help from other developers? [Add your repositories](https://issuehunt.io/r/new) on Issuehunt to raise funds.

### PR DESCRIPTION
**Pull request by GMartigny**
Fix #157

### CODE REVIEW

File: rules/test-title-format.js [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+
+const create = context => {
+	const ava = createAvaRule();
+	let titleRegExp;
+	try {
+		if (context.options[0] && context.options[0].regexp) {
+			titleRegExp = new RegExp(context.options[0].regexp);
+		}
+	} catch (error) {
+		console.warn(error);
```
Comment (sindresorhus) [inline rules/test-title-format.js]: I would use `context.report` for this too, so it includes the file. And then return early. Just set the node to `0:0`. Unless there's a better ESLint API to use for reports like this?

File: rules/test-title-format.js [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+	return ava.merge({
+		CallExpression: visitIf([
+			ava.isInTestFile,
+			ava.isTestNode,
+			ava.hasNoHookModifier
+		])(node => {
+			const requiredLength = ava.hasTestModifier('todo') ? 1 : 2;
+			const hasTitle = node.arguments.length >= requiredLength;
+
+			if (hasTitle && titleRegExp) {
```
Comment (sindresorhus) [inline rules/test-title-format.js]: `titleRegExp` should be handled by the early return.

File: rules/test-title-format.js [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+				const title = node.arguments[0];
+				if (title.type === 'Literal' && !titleRegExp.test(title.value)) {
+					context.report({
+						node,
+						message: 'Title doesn\'t match required format.'
```
Comment (sindresorhus) [inline rules/test-title-format.js]:
```suggestion
						message: 'The test title doesn\'t match the required format: '
```

And include the regex stringified.

**Comment by sindresorhus** [comment]
You also need to add it to the readme.

**Comment by sindresorhus** [comment]
Can you help review the other PRs? 🙏

File: docs/rules/test-title-format.md [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+	t.pass();
+});
+```
+
+
+## Options
+
+This rule supports the following options:
+
+`regexp`: A regular expression string to match against the test titles. Overrides the default and the configuration found in the `package.json` or `ava.config.js` files.
```
[the remark this answers is not in the document]
Comment (GMartigny) [reply docs/rules/test-title-format.md]: It's a lazy copy/paste from the `no-import-test-files`. Should I just remove this ?

[the remark this answers is not in the document]
Comment (GMartigny) [reply docs/rules/test-title-format.md]: I'll go with `format` since it match with the rule's name.

[the remark this answers is not in the document]
Comment (sindresorhus) [reply docs/rules/test-title-format.md]: Yes

File: rules/test-title-format.js (hunk 1 above)
Comment (GMartigny) [reply rules/test-title-format.js]: I couldn't find a better API. Other eslint-plugins don't even bother try-catch: 
 * https://github.com/yannickcr/eslint-plugin-react/blob/6e4f43ef4739191a2157659de178d1cd9a28a42e/lib/rules/boolean-prop-naming.js#L57
 * https://github.com/benmosher/eslint-plugin-import/blob/70a59fe1880b4061cd7c0f032c0b05c728223a0f/utils/moduleVisitor.js#L19

Comment (sindresorhus) [reply rules/test-title-format.js]: Yeah, same with ESLint. I think it's better to just drop the try/catch then and let it crash if invalid regex.

**Comment by GMartigny** [comment]
> Can you help review the other PRs?

Yes, I don't know up to what point I'm relevant, but I'll try to help.

**Comment by GMartigny** [comment]
Also, I have no idea how to (or even if I should) test crashing test-case. Maybe by supercharging @jfmengels's [`eslint-ava-rule-tester`](https://github.com/jfmengels/eslint-ava-rule-tester) ?

**Comment by sindresorhus** [comment]
> Also, I have no idea how to (or even if I should) test crashing test-case.

What do you mean by crashing? No rule should be crashing, so if it is, it should be fixed and then adding a normal (regression) test.

File: docs/rules/test-title-format.md [hunk 5]
```diff
@@ -0,0 +1,67 @@
+# Ensure tests have a correct title format
```
Comment (sindresorhus) [inline docs/rules/test-title-format.md]:
```suggestion
# Enforce test titles have a certain format
```

File: test/test-title-format.js [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+			errors
+		},
+		{
+			code: header + 'test.todo("Test something");',
+			options: [{format: '^Should'}],
+			errors
+		},
+		{
+			code: header + 'test("Foo", t => { t.pass(); });',
+			options: [{format: '(]'}], // Invalid regexp
```
Comment (sindresorhus) [inline test/test-title-format.js]:
```suggestion
			options: [{format: '(]'}], // Invalid RegExp
```

File: docs/rules/test-title-format.md [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+
+Tests should have a title matching the format option.
```
Comment (sindresorhus) [inline docs/rules/test-title-format.md]: I think we can just drop this line. It doesn't add much over the title. Instead, would be good if you could mention some use-cases for the rule.

**Comment by GMartigny** [comment]
> No rule should be crashing

I was talking about the case of an invalid regexp (https://github.com/avajs/eslint-plugin-ava/pull/235#discussion_r279691625)

**Comment by sindresorhus** [comment]
I don't think we need to test that.

File: docs/rules/test-title-format.md [hunk 8]
```diff
@@ -0,0 +1,69 @@
+# Enforce test titles have a certain format
```
Comment (sindresorhus) [inline docs/rules/test-title-format.md]:
```suggestion
# Ensure test titles have a certain format
```

And don't forget to update the description in the readme too.

**Comment by sindresorhus** [comment]
Can you fix the merge conflict?

**Comment by sindresorhus** [comment]
This was not done properly: https://github.com/avajs/eslint-plugin-ava/pull/235#discussion_r285338488

File: rules/test-title-format.js [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+		])(node => {
+			const requiredLength = ava.hasTestModifier('todo') ? 1 : 2;
+			const hasTitle = node.arguments.length >= requiredLength;
+
+			if (hasTitle) {
+				const title = node.arguments[0];
+				if (title.type === 'Literal' && !titleRegExp.test(title.value)) {
+					context.report({
+						node,
+						message: `The test title doesn't match the required format: ${titleRegExp}`
```
Comment (sindresorhus) [inline rules/test-title-format.js]:
```suggestion
						message: `The test title doesn't match the required format: \`${titleRegExp}\`.`
```

**Comment by sindresorhus** [comment]
> https://github.com/avajs/eslint-plugin-ava/pull/235#issuecomment-495079649

⬆️

**Comment by GMartigny** [comment]
> This was not done properly: [#235 (comment)](https://github.com/avajs/eslint-plugin-ava/pull/235#discussion_r285338488)

Sorry, I thought you were talking about the `Enforce` -> `Ensure` bit. All good now.

**Pull request merged**